### PR TITLE
LUCENE-9616: Add developer docs on how to update a format.

### PR DESCRIPTION
**Pull request by jtibshirani**
This commit adds simple guidelines on how to make a change to a file format:
* Document how the 'copy-on-write' approach works with backwards-codecs
* Clarify that we prefer to copy the format instead of using internal versions

### CODE REVIEW

**Comment by jtibshirani** [comment]
I'm not sure this is the right place to add developer docs... let me know if a wiki page would be more appropriate!

File: lucene/backward-codecs/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+used for reading, then delete the write-side logic and move it to a test-only
+class like Lucene80RWNormsFormat to support unit tests. Note that most formats
+only need read logic, but a small set including DocValuesFormat and
+FieldInfosFormat will need to retain write logic since can be used to update
+old segments.
+3. Make a change to the new format!
+
+## Internal format versions
+
+Each format class maintains an internal version which is written into the
```
Comment (jtibshirani) [inline lucene/backward-codecs/README.md:40]: This is based on the discussion in https://issues.apache.org/jira/browse/LUCENE-9616.

File: lucene/backward-codecs/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+Codecs and file formats are versioned according to the minor version in which
+they were created. For example Lucene87Codec represents the codec used for
+creating Lucene 8.7 indices, and potentially later index versions too. Each
+segment records the codec version that was used to write it.
+
+Lucene supports the ability to read segments created in older versions by
+maintaining old codec classes. These older codecs live in the backwards-codecs
+package along with their file formats. When making a change to a file format,
+we create a fresh copies of the codec and format, and move the existing ones
```
Comment (msokolov) [inline lucene/backward-codecs/README.md:16]: "we create fresh copies" (strike "a")

File: lucene/backward-codecs/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+1. Create a new format with the target version for the changes, for example
+Lucene90NormsFormat. This includes creating copies of its writer and reader
+classes, as well as any helper classes. Make sure to copy unit tests too, like
+TestLucene80NormsFormat.
+2. Move the old Lucene80NormsFormat, along with its writer, reader, tests, and
+helper classes to the backwards-codecs package. If the format will only be
+used for reading, then delete the write-side logic and move it to a test-only
+class like Lucene80RWNormsFormat to support unit tests. Note that most formats
+only need read logic, but a small set including DocValuesFormat and
+FieldInfosFormat will need to retain write logic since can be used to update
```
Comment (msokolov) [inline lucene/backward-codecs/README.md:40]: "since they can be used"

File: lucene/backward-codecs/README.md (hunk 1 above)
Comment (msokolov) [reply lucene/backward-codecs/README.md:49]: Maybe in package-level javadocs: https://lucene.apache.org/core/8_8_0/backward-codecs/index.html

I would vote for either what you have done here, or javadocs, in preference to wiki

Comment (jtibshirani) [reply lucene/backward-codecs/README.md:49]: Thanks for the copy-edits. I wasn't sure if javadocs were the best place as they tend to be user-facing (describing how to use backwards-codecs, for example, as opposed to the development details).

File: lucene/backward-codecs/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+This README describes the approach to maintaining compatibility with indices
+from previous versions and gives guidelines for making format changes.
+
+## Compatibility strategy
+
+Codecs and file formats are versioned according to the minor version in which
+they were created. For example Lucene87Codec represents the codec used for
+creating Lucene 8.7 indices, and potentially later index versions too. Each
+segment records the codec version that was used to write it.
```
Comment (jpountz) [inline lucene/backward-codecs/README.md:11]: I think that the following would be more accurate?

```suggestion
segment records the codec name that was used to write it.
```

**Review by jpountz — APPROVED** [review]

**Pull request merged**